Support CSharp Light Bulbs (FQN, Add Using)

TanayParikh opened the pull request:
### Summary of the changes

- Support for Fully Qualified Namespace C# Light Bulb
![FullyQualifiedCodeAction](https://user-images.githubusercontent.com/14852843/93262976-f1da7f80-f759-11ea-9877-05e9df2ed88b.gif)

- Support for Add Using C# Light bulb
![AddUsingCodeAction](https://user-images.githubusercontent.com/14852843/93263130-2d754980-f75a-11ea-890d-bfb07516a354.gif)

- Abstraction Layer / Refactoring to create shared light bulb infrastructure between TSLanguageServer + HTMLCSharpLanguageServer.
- This implementation doesn't have support for code actions which require resolving with Roslyn (ex. via `textDocument/codeActionResolve`). Working on an appropriate expansion for this, and will likely require changes in `omnisharp-vscode`.
    - Tracked here: https://github.com/dotnet/aspnetcore/issues/25918
- Some general perf optimizations are possible, but wanted to avoid potential pre-optimizations. Will re-visit if we see issues.
- Testing: TODO assuming no issues with the design.

Fixes: https://github.com/dotnet/aspnetcore/issues/24779
Fixes: https://github.com/dotnet/aspnetcore/issues/18173
Fixes: https://github.com/dotnet/aspnetcore/issues/24778
Fixes: https://github.com/dotnet/aspnetcore/issues/25144

TanayParikh commented on src/Razor/src/Microsoft.VisualStudio.LanguageServerClient.Razor/DefaultRazorLanguageServerCustomMessageTarget.cs:
```diff
@@ -166,5 +167,35 @@ public override async Task<RazorDocumentRangeFormattingResponse> RazorRangeForma
+                throw new ArgumentNullException(nameof(codeActionParams));
+            }
+
+            if (!_documentManager.TryGetDocument(codeActionParams.TextDocument.Uri, out var documentSnapshot))
+            {
+                return default;
+            }
+
+            cancellationToken.ThrowIfCancellationRequested();
+
+            if (!documentSnapshot.TryGetVirtualDocument<CSharpVirtualDocumentSnapshot>(out var csharpDoc))
+            {
+                return default;
+            }
```
This is done just to get the generated C# URI from the `.razor` file path.

TanayParikh commented on src/Razor/src/Microsoft.AspNetCore.Razor.VSCode/src/CodeActions/CodeActionsHandler.ts:
```diff
@@ -0,0 +1,71 @@
+    }
+
+    private commandAsCodeAction(command: vscode.Command): RazorCodeAction {
+        return { title: command.title } as RazorCodeAction;
```
This will be expanded with the code action resolve implementation.

TanayParikh commented on src/Razor/src/Microsoft.AspNetCore.Razor.LanguageServer/CodeActions/AddUsingsCodeActionProviderHelper.cs:
```diff
@@ -0,0 +1,54 @@
+
+namespace Microsoft.AspNetCore.Razor.LanguageServer.CodeActions
+{
+    internal static class AddUsingsCodeActionProviderHelper
```
Helper shared between C# & Razor add usings.

TanayParikh commented on src/Razor/src/Microsoft.VisualStudio.LanguageServerClient.Razor/CSharpVirtualDocumentPublisher.cs:
```diff
@@ -70,11 +70,7 @@ private class CSharpVirtualDocumentContainer : DynamicDocumentContainer
-            {
-                get => true;
-                set { /* no-op */ }
-            }
```
Unrelated change from earlier PR feedback.

TanayParikh commented on src/Razor/test/Microsoft.VisualStudio.LanguageServer.ContainedLanguage.Test/DefaultLSPDocumentSynchronizerTest.cs:
```diff
@@ -26,6 +26,7 @@ public DefaultLSPDocumentSynchronizerTest()
 
         private ITextBuffer VirtualDocumentTextBuffer { get; }
 
+#pragma warning disable VSTHRD200 // Use "Async" suffix for async methods
```
Unrelated refactorings to avoid warnings.

NTaylorMullen replied:
Instead of adding this warning here ignore it at the csproj level (we already have others listed). We don't care about these warnings at all for test files.

TanayParikh commented on src/Razor/test/Microsoft.AspNetCore.Razor.LanguageServer.Test/CodeActions/CodeActionEndpointTest.cs:
```diff
@@ -24,6 +24,7 @@ namespace Microsoft.AspNetCore.Razor.LanguageServer.Test.CodeActions
 {
     public class CodeActionEndpointTest : LanguageServerTestBase
     {
+        private readonly RazorDocumentMappingService DocumentMappingService = Mock.Of<RazorDocumentMappingService>();
```
Tests updated to a passing state. Testing of functionality introduced with this PR is to be done.

I'd like to get this merged first to ensure the PR isn't too large. I'll subsequently implement the resolve endpoint, and do the testing for it all together because there's likely to be a good deal of overlap.

ajaybhargavb commented on src/Razor/src/Microsoft.AspNetCore.Razor.LanguageServer/CodeActions/AddUsingsCodeActionProviderHelper.cs:
```diff
@@ -0,0 +1,54 @@
+            return namespaceName;
+        }
+
+        internal static RazorCodeAction CreateAddUsingCodeAction(string fullyQualifiedName, DocumentUri uri)
```
Can this functionality be in a base class instead of a helper class? We generally try to avoid helper methods if possible to keep it contained to where it is intended to be used.

TanayParikh replied:
It's used in two separate providers `RazorAddUsingsCodeActionProvider` which implements `RazorCodeActionProvider` and `CSharpAddUsingsCodeActionProvider` which implements `CSharpCodeActionProvider`. So I don't think a base class would be possible without multiple-inheritance here.

ajaybhargavb commented on src/Razor/src/Microsoft.AspNetCore.Razor.LanguageServer/CodeActions/CodeActionEndpoint.cs:
```diff
@@ -74,22 +80,50 @@ public async Task<CommandOrCodeActionContainer> Handle(CodeActionParams request,
+
+            if (cancellationToken.IsCancellationRequested)
+            {
+               return null;
```
Why return null instead of throwing a task canceled exception?

TanayParikh replied:
Probably best to do task cancelled. Wanted to return null to avoid unhandled task cancellation exceptions causing external issues, but that really shouldn't be necessary.

ajaybhargavb commented on src/Razor/src/Microsoft.AspNetCore.Razor.LanguageServer/CodeActions/CodeActionEndpoint.cs:
```diff
@@ -108,25 +142,89 @@ private async Task<IEnumerable<RazorCodeAction>> GetCodeActionsAsync(CodeActionP
+
+            context.Request.Range = projectedRange;
+
             if (cancellationToken.IsCancellationRequested)
```
I guess it was like this in other places too. I'm curious why.

TanayParikh replied:
https://github.com/dotnet/aspnetcore-tooling/pull/2501#discussion_r489011372

ajaybhargavb reviewed (commented):
LGTM from a quick skim. I'll look at it more in some time.

NTaylorMullen commented on src/Razor/src/Microsoft.AspNetCore.Razor.LanguageServer/CodeActions/CodeActionEndpoint.cs:
```diff
@@ -108,25 +142,89 @@ private async Task<IEnumerable<RazorCodeAction>> GetCodeActionsAsync(CodeActionP
+            return filteredCSharpCodeActions;
+        }
+
+        private async Task<IEnumerable<RazorCodeAction>> FilterCSharpCodeActionsAsync(RazorCodeActionContext context, IEnumerable<RazorCodeAction> csharpCodeActions, CancellationToken cancellationToken)
```
Is it filter or is it "remap"?

TanayParikh replied:
Filter b/c we take the full set of C# code actions supplied by Roslyn, and filter out the ones we support in Razor.

NTaylorMullen commented on src/Razor/src/Microsoft.AspNetCore.Razor.LanguageServer/CodeActions/CSharp/CSharpCodeActionProvider.cs:
```diff
@@ -0,0 +1,15 @@
+{
+    internal abstract class CSharpCodeActionProvider
+    {
+        public abstract Task<RazorCodeAction[]> ProvideAsync(RazorCodeActionContext context, IEnumerable<RazorCodeAction> codeActions, CancellationToken cancellationToken);
```
So the gist of how this functions now is it takes in a code action context, a list of pre-existing csharp code actions and then remaps/translates them to be Razor code actions right? I wonder if "provide" is the right terminology for how this works. Also i'd suggest making the return value immutable (`IReadOnlyList<RazorCodeAction>`)

TanayParikh replied:
> and then remaps/translates them to be Razor code actions right
First ensures that specific provider can handle the code action (thereby filtering out code actions which aren't supported by razor) and then does the remap/translate if the code action is supported.

> I wonder if "provide" is the right terminology for how this works.
I kept it as `provide` to keep consistent with the `RazorCodeActionProvider` which is similar except of course it doesn't take a code action. We could do `ProcessAsync`, but we would have to change the `CSharpCodeActionProvider` name as well then. Open to suggestions

> Also i'd suggest making the return value immutable (IReadOnlyList<RazorCodeAction>)
Done.

NTaylorMullen replied:
Can leave as is for now.

NTaylorMullen commented on src/Razor/src/Microsoft.AspNetCore.Razor.LanguageServer/CodeActions/CodeActionEndpoint.cs:
```diff
@@ -108,25 +142,89 @@ private async Task<IEnumerable<RazorCodeAction>> GetCodeActionsAsync(CodeActionP
+                context.Request.Range.Start.Line,
+                context.Request.Range.Start.Character);
+            var hostDocumentIndex = context.SourceText.Lines.GetPosition(linePosition);
+            context.Location = new SourceLocation(
```
Why not do this when the context is constructed? Having the location be settable seems odd only for this.

TanayParikh replied:
There was a reason. Replying to this comment, I realize that reasoning was invalid. Fixing 😄 

We only needed location for `RazorCodeAction` so I didn't want to compute it for both, but considering we generate the context only once for both, there wouldn't be any recomputation.

NTaylorMullen commented on src/Razor/src/Microsoft.AspNetCore.Razor.LanguageServer/CodeActions/RazorCodeActionContext.cs:
```diff
@@ -14,20 +15,21 @@ internal sealed class RazorCodeActionContext
-        public SourceLocation Location { get; }
+        public SourceText SourceText { get; }
         public bool SupportsFileCreation { get; }
+        public SourceLocation Location { get; set; }
```
Feels odd for this to be settable, left a comment above where it's used.

TanayParikh replied:
Addressing comment above.

NTaylorMullen commented on src/Razor/src/Microsoft.AspNetCore.Razor.VSCode/src/CodeActions/CodeActionsHandler.ts:
```diff
@@ -0,0 +1,71 @@
+        try {
+            const razorDocumentUri = vscode.Uri.parse(codeActionParams.textDocument.uri);
+            const razorDocument = await this.documentManager.getDocument(razorDocumentUri);
+            const virtualCsharpUri = razorDocument.csharpDocument.uri;
```
Should validate that the `razorDocument` is available before doing this. If it was closed or deleted it'll be `null` here.

TanayParikh replied:
Added `undefined` check.

NTaylorMullen commented on src/Razor/src/Microsoft.AspNetCore.Razor.VSCode/src/CodeActions/CodeActionsHandler.ts:
```diff
@@ -0,0 +1,71 @@
+
+            const commands = await vscode.commands.executeCommand<vscode.Command[]>(
+                'vscode.executeCodeActionProvider',
+                virtualCsharpUri,
```
```suggestion
                virtualCSharpUri,
```

NTaylorMullen commented on src/Razor/src/Microsoft.AspNetCore.Razor.VSCode/src/RazorCSharpLanguageMiddleware.ts:
```diff
@@ -149,12 +134,4 @@ export class RazorCSharpLanguageMiddleware implements LanguageMiddleware {
         }
     }
-
-    private tryApplyingCodeActions(uri: vscode.Uri, edit: vscode.TextEdit): [ vscode.Uri?, vscode.TextEdit?] {
```
Really love all the cleanup

NTaylorMullen commented on src/Razor/src/Microsoft.AspNetCore.Razor.VSCode/src/CodeActions/CodeActionsHandler.ts:
```diff
@@ -0,0 +1,71 @@
+import { convertRangeFromSerializable } from '../RPC/SerializableRange';
+
+export class CodeActionsHandler
+    extends RazorLanguageFeatureBase {
```
This feature base is only relevant for features that call `getProjectionAsync`. In your case you don't so I don't think you need all of the ceremony/requirements this base class requires

TanayParikh replied:
Yep thanks. Removed the need for `getProjectionAsync` but missed removing the base class.

NTaylorMullen commented on src/Razor/src/Microsoft.CodeAnalysis.Razor.Workspaces/DynamicDocumentContainer.cs:
```diff
@@ -9,7 +9,7 @@ internal abstract class DynamicDocumentContainer
         public abstract string FilePath { get; }
 
-        public abstract bool SupportsDiagnostics { get; set; }
+        public virtual bool SupportsDiagnostics { get; set; }
```
Why is abstract not ok?

TanayParikh replied:
This was a little while back, but I believe it was so we could override the getter without having to `no-op` the setter.
https://github.com/dotnet/aspnetcore-tooling/pull/2501#discussion_r488973199

NTaylorMullen replied:
Hmm, feels odd but if you want to go with it we can try it out

NTaylorMullen commented on src/Razor/src/Microsoft.VisualStudio.LanguageServerClient.Razor/DefaultRazorLanguageServerCustomMessageTarget.cs:
```diff
@@ -166,5 +167,35 @@ public override async Task<RazorDocumentRangeFormattingResponse> RazorRangeForma
+                return default;
+            }
+
+            cancellationToken.ThrowIfCancellationRequested();
```
Nothing's async, I wouldn't add this check here

NTaylorMullen commented on src/Razor/src/Microsoft.VisualStudio.LanguageServerClient.Razor/RazorLanguageServerCustomMessageTarget.cs:
```diff
@@ -22,5 +24,9 @@ internal abstract class RazorLanguageServerCustomMessageTarget
         [JsonRpcMethod(LanguageServerConstants.RazorRangeFormattingEndpoint, UseSingleObjectParameterDeserialization = true)]
         public abstract Task<RazorDocumentRangeFormattingResponse> RazorRangeFormattingAsync(RazorDocumentRangeFormattingParams token, CancellationToken cancellationToken);
+
+        // Called by the Razor Language Server to provide code actions from the platform.
```
👍

NTaylorMullen reviewed (changes requested):
Hot damn this is awesome to see land! 👏 

Love where you're going with it. Had some suggestions and a few questions throughout.

TanayParikh commented on src/Razor/src/Microsoft.AspNetCore.Razor.LanguageServer/CodeActions/CSharp/TypeAccessibilityCodeActionProvider.cs:
```diff
@@ -0,0 +1,103 @@
+
+        public override Task<RazorCodeAction[]> ProvideAsync(RazorCodeActionContext context, IEnumerable<RazorCodeAction> codeActions, CancellationToken cancellationToken)
+        {
+            var diagnostic = context.Request.Context.Diagnostics.FirstOrDefault(diagnostic =>
```
> ambiguous references where two or more namespaces are applicable

Hadn't considered that case, thanks! Updated.

TanayParikh commented on src/Razor/src/Microsoft.AspNetCore.Razor.LanguageServer/CodeActions/CSharp/TypeAccessibilityCodeActionProvider.cs:
```diff
@@ -0,0 +1,103 @@
+            return Task.FromResult(results.ToArray());
+        }
+
+        private static void CreateFQNCodeAction(
```
Thanks, that's a great point. Updated.

TanayParikh commented on src/Razor/src/Microsoft.AspNetCore.Razor.LanguageServer/CodeActions/CSharp/TypeAccessibilityCodeActionProvider.cs:
```diff
@@ -0,0 +1,103 @@
+            RazorCodeAction codeAction,
+            ICollection<RazorCodeAction> results)
+        {
+            var codeDocumentIdentifier = new VersionedTextDocumentIdentifier() { Uri = context.Request.TextDocument.Uri };
```
This will require changes throughout Code Actions (& beyond?), created https://github.com/dotnet/aspnetcore/issues/25936 to address separately.

TanayParikh commented on src/Razor/src/Microsoft.AspNetCore.Razor.LanguageServer/CodeActions/AddUsingsCodeActionProviderHelper.cs:
```diff
@@ -0,0 +1,54 @@
+{
+    internal static class AddUsingsCodeActionProviderHelper
+    {
+        internal static string GetNamespaceFromFQN(string fullyQualifiedName)
```
> Are you sure translating a FQN into an Add using is always correct 

No.

> understanding when C# provides the using light bulb?

So this was the alternative. It seemed the solution here would likely add a fair bit of complexity, given we have to make changes to how we handle the code behind as @ryanbrandenburg detailed [here](https://github.com/dotnet/aspnetcore/issues/18173). The TLDR is we have `#line hidden`s around the `using` block at the start of the generated document, as well as some near the code block area. This results in Roslyn not providing any `using` code action for us. 

I'm not sure how much of an issue this would present in practice / whether it's worth prioritizing over other changes / code actions. Thoughts?

> If the assumption is wrong we could potentially be offering the add using light bulb when the using statement already exists and invoking the light bulb would result in the Razor document being invalid.

If we re-add a using statement that already exists, it shouldn't matter(?) as the compiler just ignores duplicate `using`. Also, as this is just doing a `@using` at the top of the razor, unexpected behavior can quickly be rectified with a undo (no corruption of background document or anything).

NTaylorMullen replied:
> I'm not sure how much of an issue this would present in practice / whether it's worth prioritizing over other changes / code actions. Thoughts?

To be honest I totally forgot about the Roslyn implication of them not actually providing the using light bulb due to our hidden directives. Because of that what you have here is an awesome alternative. I think eventually we'll probably want to work with Roslyn to enable those light bulbs in Razor files despite the line directives but there's a lot of value in seeing how far the solution you have here can get users. Definitely something we can re-evaluate in the future based on feedback.



> If we re-add a using statement that already exists, it shouldn't matter(?) as the compiler just ignores duplicate `using`. Also, as this is just doing a `@using` at the top of the razor, unexpected behavior can quickly be rectified with a undo (no corruption of background document or anything).

Technically it shouldn't matter; however, if adding that using statement then makes other pieces of the document "ambiguous" adding the using statement could cause errors. That being said, it's not a common case scenario so it's definitely worth ignoring for the time being.

TanayParikh commented on src/Razor/src/Microsoft.AspNetCore.Razor.LanguageServer/CodeActions/CSharp/TypeAccessibilityCodeActionProvider.cs:
```diff
@@ -0,0 +1,103 @@
+
+            foreach (var codeAction in codeActions)
+            {
+                if (!codeAction.Title.Any(c => char.IsWhiteSpace(c)) &&
```
Created: https://github.com/dotnet/aspnetcore/issues/25939 to track the effort.

TanayParikh commented:
Thanks @NTaylorMullen @ajaybhargavb. Addressed the feedback.

Noticing some issues where fully qualifying `<Hello>` results in `<SomeComponent.Hello.Hello>` despite the preview showing `<SomeComponent.Hello>`. Considering the preview is correct, I believe this is an external regression.

I'm also not getting any code actions with `CTRL + .` (shows no code actions found), but I do get them when I right click and select quick actions. I believe this is likely also an external regression. 

Is anyone else seeing this on latest VS?

NTaylorMullen reviewed (approved):
Poifect!! Looks good to me!

ajaybhargavb commented on src/Razor/src/Microsoft.AspNetCore.Razor.LanguageServer/CodeActions/Razor/ExtractToCodeBehindCodeActionProvider.cs:
```diff
@@ -13,16 +13,17 @@
 using Microsoft.AspNetCore.Razor.Language.Legacy;
 using Microsoft.AspNetCore.Razor.LanguageServer.CodeActions.Models;
 using Microsoft.AspNetCore.Razor.LanguageServer.Common;
+using System.Collections.Generic;
```
Sort

ajaybhargavb reviewed (approved):
🎉
